Close response InputStreams

poketim opened the pull request:
Closes #848 

#### What has been done to verify that this works as intended?

Code complies, all tests pass, and application runs.

#### Why is this the best possible solution? Were any other approaches considered?

As discussed in the issue, closing the InputStreams after we received the response will help to reduce the number of files opened during a large number of submissions or attachments.

Other approaches were thought of, but would require more re-work/refactoring of existing functionality, where this current solution fits in line with how the code is organized and utilizes InputStreams from the class UncheckedFiles.

I attempted to update and refactor CentralServer using the similar closeInputStream method, but ran into 3 failing unit tests. Code can be found here: https://github.com/timkoon/briefcase/commit/915e262d78ef2e1a32ceb3e3bfedb9346798a0d9
That change may be verified in a separate pull request as issue 848 just references AggregateServer.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

The changes should reduce the errors regarding too many files opened. There shouldn't be other impacts to a user's experience. As a note, after the PushToAggregate operation receives a response, it now does some cleaning up (closing input streams and clearing the tracked streams opened).

#### Does this change require updates to documentation? If so, please file an issue at https://github.com/opendatakit/docs/issues/new and include the link below.

I don't believe so.

poketim committed:
Created and implemented closeInputStream method. Fix #848

poketim committed:
refactored aggregate server with closeInputStream

timkoon commented:
@getodk/briefcase

lognaturel commented:
Thank you so much for taking this on, @timkoon, and sorry about the radio silence!

I am unfortunately not very familiar with this code. One thing that I've been wanting to verify is that it indeed is necessary to keep track of all streams and close them all at once at the end. Did you verify that the streams are indeed still open at that point? I would have expected the methods in RequestBuilder at https://github.com/getodk/briefcase/blob/7bb6185e6c3212e9dbe791df32fde25b21ab4a29/src/org/opendatakit/briefcase/reused/http/RequestBuilder.java#L111 and below to close them by using try with resources.

ggalmazor commented on src/org/opendatakit/briefcase/push/aggregate/PushToAggregate.java:
```diff
@@ -164,7 +166,13 @@ void pushFormAndAttachments(FormStatus form, List<Path> attachments, RunnerStatu
-    Response response = http.execute(server.getPushFormRequest(form.getFormFile(briefcaseDir), attachments));
+    Response<InputStream> response = http.execute(server.getPushFormRequest(form.getFormFile(briefcaseDir), attachments));
+
+    for (InputStream stream : server.getFileStreams()) {
+      closeInputStream(stream);
+    }
```
```suggestion
    server.getFileStreams().forEach(UncheckedFiles::closeInputStream);
```

ggalmazor commented on src/org/opendatakit/briefcase/push/aggregate/PushToAggregate.java:
```diff
@@ -186,6 +194,12 @@ void pushSubmissionAndAttachments(Path submissionFile, List<Path> attachments, R
         attachments
     ));
+
+    for (InputStream stream : server.getFileStreams()) {
+      closeInputStream(stream);
+    }
```
```suggestion
    server.getFileStreams().forEach(UncheckedFiles::closeInputStream);
```

ggalmazor commented on src/org/opendatakit/briefcase/reused/UncheckedFiles.java:
```diff
@@ -301,6 +301,14 @@ public static InputStream newInputStream(Path path, OpenOption... options) {
     }
   }
 
+  public static void closeInputStream(InputStream inputStream) {
```
Loving this!

ggalmazor commented on src/org/opendatakit/briefcase/reused/transfer/AggregateServer.java:
```diff
@@ -54,10 +55,12 @@
 public class AggregateServer implements RemoteServer {
   private final URL baseUrl;
   private final Optional<Credentials> credentials;
+  private List<InputStream> fileStreams;
 
   public AggregateServer(URL baseUrl, Optional<Credentials> credentials) {
     this.baseUrl = baseUrl;
     this.credentials = credentials;
+    this.fileStreams = new ArrayList<>();
```
```suggestion
  private final List<InputStream> fileStreams = new ArrayList<>();

  public AggregateServer(URL baseUrl, Optional<Credentials> credentials, Array) {
    this.baseUrl = baseUrl;
    this.credentials = credentials;
```

ggalmazor replied:
I suggest making the list `final`, since we're not really want to support reassigning a new value. Also, since this is a property belonging to the internal workings of the class, I think it's OK to assign its value directly in the property declaration.

ggalmazor reviewed (approved):
This is looking good to me, although I have some reservations about making the server class aware of the file streams that some of its requests open. While I believe this alleviates the original issue, I think this should be something that the `Request` should provide instead of the server, along the lines of "file resources opened by this request" so that the HTTP client can close them automatically once the request is sent.

Having said that, I think this doesn't block the release of this PR, although I would consider opening a new PR to address the design improvements I've mentioned above.

I'm also missing tests for this change, which would be nice to have.

poketim committed:
Merge branch 'master' into issue-848

poketim committed:
Implemented review suggestions

poketim committed:
Reverted previous fix & implemented alternate fix

poketim committed:
Fixed checkstyle issues with tests

timkoon commented:
> Thank you so much for taking this on, @timkoon, and sorry about the radio silence!
> 
> I am unfortunately not very familiar with this code. One thing that I've been wanting to verify is that it indeed is necessary to keep track of all streams and close them all at once at the end. Did you verify that the streams are indeed still open at that point? I would have expected the methods in RequestBuilder at
> 
> https://github.com/getodk/briefcase/blob/7bb6185e6c3212e9dbe791df32fde25b21ab4a29/src/org/opendatakit/briefcase/reused/http/RequestBuilder.java#L111
> 
> and below to close them by using try with resources. Or is it `Request.body` that never gets closed?

Hi @lognaturel and no problem! In regards to pointing out the line in the RequestBuilder, I would think that approach is optimal when it comes to attachments in the request, but attachments in the request object are actually using `UncheckedFiles.newInputStream`. That method isn't using try-with-resources, so the InputStreams created will need to be closed and are causing the exception in #848 .

Here is where I found Aggregate server is calling that static method:
https://github.com/getodk/briefcase/blob/7bb6185e6c3212e9dbe791df32fde25b21ab4a29/src/org/opendatakit/briefcase/reused/transfer/AggregateServer.java#L160

I haven't verified that the InputStreams are still open while the application is running, but that could possibly be done in an integration test. I did include unit tests to verify Java InputStreams as working as expected, and that `inputStream.close()` closes and deallocates files/resources back to the system.

@ggalmazor Thanks for your review and suggestion. I do like your suggestion of making the Client aware and handle closing InputStreams rather than the Server. After reviewing the flow again, I found the client is already tracking the attachments and their InputStreams, so I would agree that your suggestion is a better alternative and implemented that in the new commits.

poketim committed:
Fixed typo in log message

lognaturel reviewed (approved):
This looks like a great improvement to me, @timkoon! It's now very clear which streams could be left open. The change feels nicely targeted and tested.

I'll give @ggalmazor a little bit in case he has comments too.

ggalmazor commented:
(Apologies! will have some cycles for this during the weekend)

ggalmazor commented on src/org/opendatakit/briefcase/reused/http/CommonsHttp.java:
```diff
@@ -39,12 +39,16 @@
   private Executor executor;
   private final int maxConnections;
   private final BasicCookieStore cookieStore;
+  private static final Logger LOG = LoggerFactory.getLogger(CommonsHttp.class);
```
I know this is a static member, but I believe we use `logger` as name for this elsewhere.

timkoon replied:
Sure, it makes sense to be consistent.

ggalmazor commented on src/org/opendatakit/briefcase/reused/http/CommonsHttp.java:
```diff
@@ -147,6 +151,16 @@ public void unsetProxy() {
+      try {
+        UncheckedFiles.closeInputStream(request.getBody());
+      } catch (final BriefcaseException exception) {
+        LOG.error("Error: Attempted to close Request.body InputStream, but failed", exception);
```
Not sure if this is something actionable or that would help us help a user having trouble with Briefcase. @timkoon, do you think it would be useful to have this on the log?

timkoon replied:
True, it may not be actionable, but in the least, it would give insight into any system issues. I think there's value for the log to be there even though it may be rare that it will trigger.

With the alternative, I don't think there's value if the catch statement does nothing.

ggalmazor commented on src/org/opendatakit/briefcase/reused/http/CommonsHttp.java:
```diff
@@ -147,6 +151,16 @@ public void unsetProxy() {
     } catch (IOException e) {
       throw new UncheckedIOException(e);
+    } finally {
+      try {
```
I wish we could use try with resources but I guess it's up to the HTTP client's implementation. I recon that this would work as we want.

timkoon replied:
I agree. There's a chance that the code can be refactored at a later point to possibly accomplish that.

ggalmazor commented on test/java/org/opendatakit/briefcase/reused/UncheckedFilesTest.java:
```diff
@@ -0,0 +1,65 @@
+    inputStream.available();
+  }
+
+  @Test(expected = UncheckedIOException.class)
+  public void closeInputStream_should_throw_exception() throws IOException {
+    InputStream inputStream = mock(InputStream.class);
+    doThrow(new IOException()).when(inputStream).close();
+
+    UncheckedFiles.closeInputStream(inputStream);
+  }
```
@Test(expected = UncheckedIOException.class)
  public void closeInputStream_should_throw_exception() throws IOException {
    InputStream is = new InputStream() {
      @Override
      public int read() throws IOException {
        return 0;
      }

      @Override
      public void close() throws IOException {
        throw new IOException("chuchu");
      }
    };
    UncheckedFiles.closeInputStream(is);
  }

ggalmazor commented on test/java/org/opendatakit/briefcase/reused/UncheckedFilesTest.java:
```diff
@@ -0,0 +1,65 @@
+    inputStream.available();
+  }
+
+  @Test(expected = UncheckedIOException.class)
+  public void closeInputStream_should_throw_exception() throws IOException {
+    InputStream inputStream = mock(InputStream.class);
+    doThrow(new IOException()).when(inputStream).close();
+
+    UncheckedFiles.closeInputStream(inputStream);
+  }
```
I don't think that this test justifies adding Mockito to the project. I know it's ugly, but the suggested change would take care of the test without Mockito.

```suggestion
  @Test(expected = UncheckedIOException.class)
  public void closeInputStream_should_throw_exception() throws IOException {
    InputStream is = new InputStream() {
      @Override
      public int read() throws IOException {
        return 0;
      }

      @Override
      public void close() throws IOException {
        throw new IOException("chuchu");
      }
    };
    UncheckedFiles.closeInputStream(is);
  }
```

timkoon replied:
Sure, my thinking was that mockito could be introduced and used in other tests to help cover exception handling or other cases. 

Maybe it can be introduced when a pull request actually contains moderate usage of mockito.

ggalmazor commented on build.gradle:
```diff
@@ -62,6 +62,7 @@ dependencies {
   testCompile 'com.github.npathai:hamcrest-optional:2.0.0'
   testCompile 'org.assertj:assertj-swing-junit:3.8.0' // Required for Swing tests
   testCompile "com.github.dreamhead:moco-core:0.12.0"
+  testCompile group: 'org.mockito', name: 'mockito-core', version: '3.4.0'
```
I think we need to remove this. See my suggestion about the test that uses this

ggalmazor commented on test/java/org/opendatakit/briefcase/reused/UncheckedFilesTest.java:
```diff
@@ -0,0 +1,65 @@
+import org.junit.Test;
+
+
+public class UncheckedFilesTest {
```
Kudos for addressing this! My only suggestion would be to change its name to `UncheckedFilesInputStreamTest`.

ggalmazor commented on test/java/org/opendatakit/briefcase/reused/UncheckedFilesTest.java:
```diff
@@ -0,0 +1,65 @@
+
+  @After
+  public void tearDown() {
+    deleteRecursive(tempDir);
```
Nice touch! ❤️

ggalmazor commented on test/java/org/opendatakit/briefcase/reused/UncheckedFilesTest.java:
```diff
@@ -0,0 +1,65 @@
+    deleteRecursive(tempDir);
+  }
+
+  @Test(expected = UncheckedIOException.class)
```
I think this test is not really necessary because then it feels like we should test all the other possible causes for `Files.newInputStream` to throw an `IOException`

ggalmazor commented on test/java/org/opendatakit/briefcase/reused/UncheckedFilesTest.java:
```diff
@@ -0,0 +1,65 @@
+  }
+
+  @Test
+  public void closeInputStream_should_handle_null() {
```
I think this test is redundant because it's testing what's effectively a NO_OP without side effects.

This IMO is telling us that we need to remove the new `if` block in `closeInputStream` that silently allows for `null` inputStreams to be provided.

ggalmazor commented on src/org/opendatakit/briefcase/reused/UncheckedFiles.java:
```diff
@@ -302,6 +302,8 @@ public static InputStream newInputStream(Path path, OpenOption... options) {
   }
 
   public static void closeInputStream(InputStream inputStream) {
+    if (inputStream == null)
```
This feels wrong. What's the rationale behind supporting a `null` `inputStream` arg?

timkoon replied:
Since it's a static method that could be called in other places, on the off chance that an inputStream is passed and is null, right now, it's not handling the null pointer exception.

The other option could be to update the catch to catch both `catch(NullPointerException | IOException error)`, and throw the UncheckedException.

Otherwise, if you believe it doesn't need to be handled, then I can remove it.

ggalmazor commented on test/java/org/opendatakit/briefcase/reused/UncheckedFilesTest.java:
```diff
@@ -0,0 +1,65 @@
+
+
+public class UncheckedFilesTest {
+  private static Path tempDir;
```
No need for this to be static because it's assigned in a `@Before` block
```suggestion
  private Path tempDir;
```

ggalmazor commented on test/java/org/opendatakit/briefcase/reused/UncheckedFilesTest.java:
```diff
@@ -0,0 +1,65 @@
+  @Before
+  public void setUp() {
+    tempDir = createTempDirectory("briefcase");
+    temp = Paths.get(tempDir.toString() + "/test.txt");
```
```suggestion
    temp = tempDir.resolve("test.txt");
```
We should use the API in `Path` to resolve new children paths.

ggalmazor commented on test/java/org/opendatakit/briefcase/reused/UncheckedFilesTest.java:
```diff
@@ -0,0 +1,65 @@
+
+  @Test(expected = IOException.class)
+  public void closeInputStream_should_close() throws IOException {
+    InputStream inputStream = UncheckedFiles.newInputStream(Paths.get(tempDir.toString() + "/test.txt"));
```
```suggestion
    InputStream inputStream = UncheckedFiles.newInputStream(temp);
```
You can use `temp` here instead of building a new `Path` (also, remember to use `Path.resolve()` ;) )

ggalmazor commented on test/java/org/opendatakit/briefcase/reused/UncheckedFilesTest.java:
```diff
@@ -0,0 +1,65 @@
+  @Test(expected = IOException.class)
+  public void closeInputStream_should_close() throws IOException {
+    InputStream inputStream = UncheckedFiles.newInputStream(Paths.get(tempDir.toString() + "/test.txt"));
+    // close inputStream
```
```suggestion
```
I think we can skip this comment

ggalmazor reviewed (changes requested):
Thanks for considering my suggestions. I think the implementation is ready, but I've seen some things in the new test class that should be addressed, so I'm requesting some changes there. Feel free to ping me anytime if you think I could help you with that ;)

poketim committed:
Merge branch 'master' into issue-848

poketim committed:
Implemented code review suggestions

timkoon commented:
@ggalmazor Do you have a chance to review the updates?

lognaturel commented:
Thanks so much for thoughtfully addressing the review feedback, @timkoon, and for the ping on this. I'm very happy with this as-is and would like to merge. @ggalmazor I can't seem to do so without your approval, though!

ggalmazor commented on src/org/opendatakit/briefcase/reused/http/CommonsHttp.java:
```diff
@@ -147,6 +151,16 @@ public void unsetProxy() {
     } catch (IOException e) {
       throw new UncheckedIOException(e);
+    } finally {
+      try {
+        UncheckedFiles.closeInputStream(request.getBody());
+      } catch (final BriefcaseException exception) {
+        log.error("Error: Attempted to close Request.body InputStream, but failed", exception);
+      }
```
We don't throw `BriefcaseException` anymore in that method.

```suggestion
      UncheckedFiles.closeInputStream(request.getBody());
```

ggalmazor replied:
Turns out that I was wrong and we do throw `BriefcaseException`, but it's when trying to access the body of an empty request. I've fixed that in a new commit.

ggalmazor commented on src/org/opendatakit/briefcase/reused/UncheckedFiles.java:
```diff
@@ -301,6 +301,16 @@ public static InputStream newInputStream(Path path, OpenOption... options) {
   }
 
+  public static void closeInputStream(InputStream inputStream) {
+    if (inputStream == null)
+      return;
```
This is breaking the existing conventions about allowing nulls.

If we ever receive `null` inputstream we should throw an error here. To avoid them, we should change the calling site, not these utility methods, which operate under narrow set of constraints.

ggalmazor commented on test/java/org/opendatakit/briefcase/reused/UncheckedFilesInputStreamTest.java:
```diff
@@ -0,0 +1,63 @@
+    deleteRecursive(tempDir);
+  }
+
+  @Test
+  public void closeInputStream_should_handle_null() {
+    UncheckedFiles.closeInputStream(null);
+  }
+
```
We shouldn't support this now.

```suggestion
```

ggalmazor commented on src/org/opendatakit/briefcase/push/aggregate/PushToAggregate.java:
```diff
@@ -21,6 +21,7 @@
 import static org.opendatakit.briefcase.reused.UncheckedFiles.list;
 import static org.opendatakit.briefcase.reused.UncheckedFiles.size;
 
+import java.io.InputStream;
```
No need for the `import` now
```suggestion
```

ggalmazor commented on src/org/opendatakit/briefcase/push/aggregate/PushToAggregate.java:
```diff
@@ -164,7 +165,8 @@ void pushFormAndAttachments(FormStatus form, List<Path> attachments, RunnerStatu
 
     tracker.trackStartSendingFormAndAttachments(part, parts);
-    Response response = http.execute(server.getPushFormRequest(form.getFormFile(briefcaseDir), attachments));
+    Response<InputStream> response = http.execute(server.getPushFormRequest(form.getFormFile(briefcaseDir), attachments));
+
```
I think these are leftover changes from a previous stage of the PR. We don't need the type parameter here anymore.
```suggestion
    Response response = http.execute(server.getPushFormRequest(form.getFormFile(briefcaseDir), attachments));
```

ggalmazor commented on src/org/opendatakit/briefcase/push/aggregate/PushToAggregate.java:
```diff
@@ -186,6 +188,7 @@ void pushSubmissionAndAttachments(Path submissionFile, List<Path> attachments, R
         submissionFile,
         attachments
     ));
+
```
```suggestion
```

ggalmazor committed:
Apply suggestions from code review

I went ahead and change some small details that I think we should really have before merging in order to be consistent with the previous code and simplify it as much as possible.

ggalmazor committed:
Prevent BriefcaseException for trying to access a request's body when it's not present

ggalmazor commented:
Hi, @timkoon, @lognaturel!

MEGA Apologies for the delay in this review. This is not how I like to do things, and I hope you can forgive me.

This PR solves the issue at hand but I think there were a couple of small things we needed to change to align the changes with previous code and make it a smidge simpler if possible. I've gone ahead and applied the changes myself to avoid delaying the PR anymore. I hope you understand.

Other than that: LGTM!

lognaturel commented:
Thanks both so much! And again, really appreciate your patience, @timkoon. Briefcase v1.18.0 is now out with this fix. 🎉 https://forum.getodk.org/t/odk-briefcase-v1-18/31039
